### ISSUE TITLE
Latency

### ISSUE THREAD

**Issue by clbarnes** [justjanne/powerline-go#129]
Out of interest, what latencies are other users seeing, for what arguments? You can find out (if you have `time` installed) by changing your `_update_ps1` function like this:

```bash
function _update_ps1() {
    echo "powerline"
    time PS1="$(your powerline-go command)"
}
```

I have powerline-go taking 100-200ms, with the command

`powerline-go -colorize-hostname -condensed -cwd-mode plain -priority 'root,cwd-path,exit,user,host,ssh,perms,git-branch,jobs,git-status' -error $?`

e.g. when `cd`ing into a directory with no python environment:

```
real	0m0.108s
user	0m0.062s
sys	0m0.051s
```

And when `cd`ing into a directory with a python environment:

```
real	0m0.191s
user	0m0.112s
sys	0m0.096s
```

It's not much in the grand scheme of things, but it's a bit of a pain given it's added to every command run! I was just wondering whether my experience was typical.

**Comment by tuananh** [comment]
yeah the delay is quite noticable when I use Windows Sublinux System. Not much so on macOS

**Comment by clbarnes** [comment]
Are you using WSL 1 or WSL 2 preview? WSL 1 has serious issues with file IO speed, among other things, so basically anything involving git is very slow.

**Comment by tuananh** [comment]
Wsl 2. Actually, on wsl 1, its very much bearable.

**Comment by tuananh** [comment]
hmm, it seems it's issue with WSL

the issue happens when i'm on /mnt

https://github.com/microsoft/WSL/issues/4197

**Comment by 82phil** [comment]
Late to the party here but found that `git stash list` was resulting in ~250ms delay for powerline while in a git repo. Using WSL1 on 1909 on Ubuntu 18.04 LTS. I have a patch applied to check if any stashes exist before pulling the list which avoids that issue.

Here are my timings
![image](https://user-images.githubusercontent.com/18205030/70126169-4e3a3b80-162d-11ea-9095-7161860a09d2.png)

I will get a pull request up sometime later for the git stash delay.

**Comment by justjanne** [comment]
The latency should be improved quite a bit now, as I changed the way git stashes and jobs are determined. If you could please test with the latest version, that’d be great :)

**Comment by 82phil** [comment]
Performance is much improved. It also doesn't regress with stashes 👍 
![image](https://user-images.githubusercontent.com/18205030/70967620-99037c80-204b-11ea-881c-d2cc679c6ed9.png)

**Comment by justjanne** [comment]
And thanks to the parallel segment PR, performance should now be even better :)

**Comment by audunmg** [comment]
It is significantly faster for me. 

Mercurial adds 250ms of latency for me, but I don't see a clean way of fixing that outside of mercurial itself

**Comment by ysiivan** [comment]
Very pronounced lag in WSL2 Ubuntu, unusable. 
Quite noticeable in PowerShell Core, but not too bad.
Fine in Linux.

**Comment by 82phil** [comment]
@ysiivan Are you using Windows or Linux directories in WSL2. There is a pretty considerable hit using the Windows side in WSL2.
![image](https://user-images.githubusercontent.com/18205030/144362299-c63c9e3e-bac7-4292-9bc0-960ce8e15e51.png)

For reference, this is how mine is configured: https://github.com/82phil/workstation/blob/main/wsl/home/.bash_additions#L7

**Comment by ysiivan** [comment]
@82phil Yes, using it on the windows directories in WSL2 is really bad. I'd be very happy with your times. However, even with 
 your config, it is way worse.
![image](https://user-images.githubusercontent.com/17600396/144499276-39ca5c2e-876e-4e34-8083-83893ddb1bf5.png)

Non windows folders 
![image](https://user-images.githubusercontent.com/17600396/144499823-32b62b0d-4c84-41f0-9a96-ce959c2bdc13.png)

For comparison, I gave a quick try to a Rust built implementation. It was faster, though likely not doing all that this one does. Still not fast enough on WSL2 windows mounted folders.

**Comment by 82phil** [comment]
@ysiivan Does this only happen in Git repos? My previous post was unintentionally misleading, those repos that I showed don't have much in them. If I use a decently sized project like Ruby on Rails I get your kind of numbers. Even using `git status` shows a large difference in performance when in the Linux file system vs Windows.
![image](https://user-images.githubusercontent.com/18205030/144541779-cee08131-ba04-41b4-9779-ff3177cb8430.png)

WSL2 file performance in Windows directories is abysmal if your churning through a lot of files. Can you move your files over to the Linux file system in WSL2?
